### ISSUE TITLE
NUnit Test Task Isn't Running

### ISSUE THREAD

**Issue by bjmarte** [cake-build/cake#158]
I'm trying to use the NUnit task to run my unit tests as part of the build process but they don't seem to be running.  I've tried to boil this down to the simplest possible scenario by hardcoding everything:

```c#
Task("Tests").Does(() =>
{
	NUnit(@"C:\_git\DTR\TestsTest\bin\debug\TestTests.dll");
});
```

I never see any test results in my output though:

```
========================================
Tests
========================================

Task                          Duration
--------------------------------------------------
Tests                         00:00:00.0260109
--------------------------------------------------
Total:                        00:00:00.0260109
```

It doesn't seem to matter if I have passing tests or failing tests, the output is the same.  I've tried running nunit-console with the exact same path and that gives meaningful output.

I know this is probably user error and not really a code issue but I couldn't find a better forum for getting help.  Thanks.

**Comment by bjmarte** [comment]
Just a few other notes...  
* Obviously that is just a snippet of my build file; I have other targets (like a compile target) that are running normally.  
* The NUnit call doesn't seem to fail no matter what crap I pass to it, even if I pass a made up file name it seems to behave the same.  
* I am using the 0.1.28 version released yesterday
* I have added Info logging lines and other things to make sure my changes are actually taking effect and the right target is getting called.

**Comment by bjmarte** [comment]
I tried using the GetFiles method on my pattern just to see how many files are found.  Either there is something wrong wtih my pattern or a hard codes path doesn't count as a valid pattern because there were no files coming back.  I changed the pattern so it started finding the right files but now I'm getting the following error:
```
========================================
Tests
========================================
Found 1 files
An error occured when executing task.
Error: NUnit: Could not locate executable.
```

I'm not sure what that error means.  I've got the NUnit location in my path and can call it from the command line.

**Comment by patriksvensson** [comment]
@bjmarte Hmm. Seems very strange since your code should work. Has this been working before and suddenly stopped working?

**Comment by patriksvensson** [comment]
@bjmarte I found the error. The globber seem to fail for actual file paths, which is kind of bad, but I've filed an issue (#159) and will try to fix it tomorrow at lunch. In the meanwhile to get your tests running, you can do like this:

```csharp
Task("Tests").Does(() =>
{
    NUnit(new FilePath[] { @"C:\_git\DTR\TestsTest\bin\debug\TestTests.dll" });
});
```

You will also need to tell Cake how to locate the NUnit runner.
There are two ways of doing this; either by putting it somewhere under the tools folder in your project root. Cake will look here by convention.

* TheProjectRoot
  * tools
    * ...
      * nunit.console.exe
  * build.cake

or by explicitly telling the Cake alias method for NUnit where it is by providing the path to it in the (normally optional) settings:

```csharp
Task("Tests").Does(() =>
{
    NUnit(new FilePath[] { @"C:\_git\DTR\TestsTest\bin\debug\TestTests.dll" },
        new NUnitSettings {
            ToolPath = "C:/The/Path/To/NUnit/Runners/nunit-console.exe"
        });
});
```

I'm sorry for this, but I'm very happy that you reported this. **Thank you!**

**Comment by bjmarte** [comment]
You are welcome.  I fixed my issue with cake not being able to find the nunit executable after digging into the cake code.  I find the tools folder convention a little strange and didn't seem to work even after I copied NUnit into the folder. It was also not easy to determine how to tell cake to look for nunit in a different location.  I'm all good now though.

**Comment by patriksvensson** [comment]
Hmm, sound strange. What did your folder structure to `nunit-console.exe` look like (from the root where you have your build script)?

The reason for the convention is that we need to have one to find things easy, and we settled on the same conventions as FAKE use for compatibility.

Sorry for the inconvenience. We're working on better documentation at the moment (better than we have at [http://cake.readthedocs.org](http://cake.readthedocs.org)), but we're not quite there yet. When the new documentation is in place, it will hopefully be easier to find information like this.

**Comment by gep13** [comment]
@patriksvensson are you planning to still use readthedocs, or are you switching to something else?

**Comment by patriksvensson** [comment]
@gep13 In a perfect world, I would like to continue to use readthedocs since it's awesome. The problem is that we will need something more complex to be able to generate proper API documentation for all public types and members of Cake.Core and Cake.Common as well as DSL documentation (for all the alias methods). This should be able to build automatically on AppVeyor when pushing to the master branch.

I've done some experimenting in the last couple of days, and I'm counting of having something in place at [http://cakebuild.net](http://cakebuild.net) in about two weeks.

**Comment by gep13** [comment]
> @patriksvensson said
I've done some experimenting in the last couple of days, and I'm counting of having something in place at http://cakebuild.net in about two weeks.

Sounds good, look forward to checking it out!

**Comment by bjmarte** [comment]
Just a couple notes for documentation sake:  

Information about how to include one script file from another would be helpful. I assumed that the syntax would be #load "<filename>", the same as scriptcs (according to their documentation, not experience).  I couldn't get that to work but then by trial and error figured out that #l "<filename>" does work.  I want to have separate files for general build, packaging, and deployments and this will be useful.  I imagine that if I have some library type functions and want to include them from multiple files that I will need to do some checking to prevent the same file from being imported twice, a documented pattern for how to accomplish that would be helpful too.

**Comment by patriksvensson** [comment]
@bjmarte You're absolutely right about the documentation. It's still a quite young project, and documentation is something that needs constant attention to be up to date. If you want to change something or just notify someone about something that's missing, the documentation can be found at (https://github.com/cake-build/documentation)[https://github.com/cake-build/documentation].

The `#l` directive will only import a script once, so you should be able to divide your scripts in separate files without problem.

**Comment by patriksvensson** [comment]
Added information about extensibility here: [http://cake.readthedocs.org/en/latest/extensibility.html](http://cake.readthedocs.org/en/latest/extensibility.html). Will add more information about addin support later.

Thanks for the feedback!

**Comment by patriksvensson** [comment]
Fixed and released in version 0.1.29.
